src/realm/db.cpp

```diff
@@ -2332,13 +2333,28 @@ TransactionRef DB::start_frozen(VersionID version_id)
 {
     if (!is_attached())
         throw LogicError(LogicError::wrong_transact_state);
+    // prune stale frozen transactions
+    std::lock_guard<std::recursive_mutex> lock(m_mutex);
+    while (!m_frozen_transactions.empty() && m_frozen_transactions.begin()->second.expired()) {
+        m_frozen_transactions.erase(m_frozen_transactions.begin());
+    }
     ReadLockInfo read_lock;
-    grab_read_lock(read_lock, version_id);
     ReadLockGuard g(*this, read_lock);
+    grab_read_lock(read_lock, version_id);
+    version_id.version = read_lock.m_version;
+    // reuse existing frozen transaction if one exists:
+    std::weak_ptr<Transaction>& weak_tr = m_frozen_transactions[version_id.version];
+    TransactionRef res = weak_tr.lock();
```

Comment:
There is likely a (small) perf benefit to spelling this `if (auto res = weak_tr.lock()) return res;`, which reduces the scope of this `res`.